File: src/typegraph/python/typegraph/runtimes/grpc.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typegraph import t
from typegraph.gen.runtimes import (
    GrpcData,
    GrpcRuntimeData,
)
from typegraph.runtimes.base import Runtime
from typegraph.sdk import runtimes


class GrpcRuntime(Runtime):
    def __init__(self, proto_file: str, endpoint: str):
        data = GrpcRuntimeData(proto_file, endpoint)
        runtime_id = runtimes.register_grpc_runtime(data)
        super().__init__(runtime_id)

    def call(self, method: str):
        data = GrpcData(method)
        func_data = runtimes.call_grpc_method(self.id, data)
        return t.func.from_type_func(func_data)

File: src/typegraph/rust/typegraph/src/runtimes/grpc.rs
use crate::{
    t::{TypeBuilder, TypeDef},
    wasm::{
        self,
        core::RuntimeId,
        runtimes::{GrpcData, GrpcRuntimeData},
    },
    Result,
};

#[derive(Debug)]
pub struct GrpcRuntime {
    id: RuntimeId,
}

impl GrpcRuntime {
    pub fn new(proto_file: &str, endpoint: &str) -> Result<Self> {
        let data = GrpcRuntimeData {
            proto_file: proto_file.to_string(),
            endpoint: endpoint.to_string(),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_grpc_runtime(s, &data))?;

        Ok(Self { id })
    }

    pub fn call(&self, method: &str) -> Result<TypeDef> {
        let data = GrpcData {
            method: method.to_string(),
        };

        wasm::with_runtimes(|r, s| r.call_call_grpc_method(s, self.id, &data))?.build()
    }
}
